### PR TITLE
Return correctly formatted column defaults in the information_schema.columns table

### PR DESCRIPTION
**Pull request by VinaiRachakonda**
The information_schema.columns contains the column `column_default` that represents the formatted string version of the column default. This correct representation is essential as tools (like table editors) use this column to generate representation of a table or their alter statements. 

This pr pushes the resolution of column defaults through the analyzer. Note that there are some open flaws

1. The double quoting of literal column defaults 
2. The double parentheses for expressions due to improper parentheses handling in the codebase
3. Missing backticks for columns. This is due to fact that all the way down to GetField for expressions we don't actually backtick the name. We should followup with a rewrite of the Column Default String method

There should be subsequent followup work that simplifies string formatting with default  values and expressions as a whole.

### CODE REVIEW

File: sql/analyzer/resolve_tables.go [hunk 1]
```diff
@@ -126,8 +128,63 @@ func resolveTable(ctx *sql.Context, t *plan.UnresolvedTable, a *Analyzer) (sql.N
 		return handleTableLookupFailure(err, name, db, a, t)
 	}
 
+	resolvedTableNode := plan.NewResolvedTable(rt, database, nil)
+
+	// Check for the information_schema.columns table which needs to resolve all defaults
+	// TODO: Is it okay to put something one off like this?
```
Comment (zachmu) [inline sql/analyzer/resolve_tables.go:134]: Yeah that's pretty much how you have to handle this

File: sql/analyzer/resolve_tables.go [hunk 2]
```diff
@@ -126,8 +128,63 @@ func resolveTable(ctx *sql.Context, t *plan.UnresolvedTable, a *Analyzer) (sql.N
...
+	ret := make(map[string]*sql.Column)
+
+	for _, db := range catalog.AllDatabases(ctx) {
+		err := sql.DBTableIter(ctx, db, func(t sql.Table) (cont bool, err error) {
+			for _, col := range t.Schema() {
+				if col.Default == nil {
+					continue
+				}
+
+				if ucd, ok := col.Default.Expression.(sql.UnresolvedColumnDefault); ok {
```
Comment (zachmu) [inline sql/analyzer/resolve_tables.go:166]: You don't need to do this here, UnresolvedColumnDefault will get resolved in resolve_column_defaults

File: sql/information_schema/columns_table.go [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+	// defaultToColumn maps the pointer of a sql.ColumnDefault value back to its original Column Object.
+	defaultToColumn map[*sql.ColumnDefaultValue]*sql.Column
+}
+
+var _ sql.Node = (*ColumnsNode)(nil)
+var _ sql.Expressioner = (*ColumnsNode)(nil)
+var _ sql.Nameable = (*ColumnsNode)(nil)
+
+// CreateNewColumnsNode returns a new ColumnsNode.
+func CreateNewColumnsNode(child sql.Node, tableToColumnsWithDefaultValue map[string]*sql.Column) *ColumnsNode {
```
Comment (zachmu) [inline sql/information_schema/columns_table.go:46]: NewColumnsTable

File: sql/information_schema/columns_table.go [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+
+	// columnNameToColumn maps the name of a column (databaseName.tableName.columnName) with a non nil default value
+	// to its column object.
+	columnNameToColumn map[string]*sql.Column
+
+	// defaultToColumn maps the pointer of a sql.ColumnDefault value back to its original Column Object.
+	defaultToColumn map[*sql.ColumnDefaultValue]*sql.Column
+}
+
+var _ sql.Node = (*ColumnsNode)(nil)
```
Comment (zachmu) [inline sql/information_schema/columns_table.go:41]: This needs to implement sql.Table in addition to sql.Node

File: sql/information_schema/columns_table.go [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
+)
+
+// ColumnsNode wraps the information_schema.columns table as a way to resolve column defaults.
+type ColumnsNode struct {
```
Comment (zachmu) [inline sql/information_schema/columns_table.go:30]: ColumnsTable

File: sql/information_schema/columns_table.go [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+}
+
+// CheckPrivileges implements the sql.Node interface.
+func (c *ColumnsNode) CheckPrivileges(ctx *sql.Context, opChecker sql.PrivilegedOperationChecker) bool {
+	return c.Child.CheckPrivileges(ctx, opChecker)
+}
+
+// Expressions implements the sql.Expressioner interface.
+func (c *ColumnsNode) Expressions() []sql.Expression {
+	c.defaultToColumn = make(map[*sql.ColumnDefaultValue]*sql.Column)
```
Comment (zachmu) [inline sql/information_schema/columns_table.go:100]: Don't use maps for any of this, it makes everything way more complicated

Just consider all the defaults for all the tables in sorted name order, and use the same order here and in WithExpressions and you'll be fine

File: sql/information_schema/information_schema.go [hunk 7]
```diff
@@ -2027,6 +1932,27 @@ func partitionKey(tableName string) []byte {
 	return []byte(InformationSchemaDatabaseName + "." + tableName)
 }
 
+// getColumnDefaultValue returns value for column default value in string format or nil for 'NULL' default value.
+func getColumnDefaultValue(cd *ColumnDefaultValue) interface{} {
```
Comment (zachmu) [inline sql/information_schema/information_schema.go:1936]: Don't do this, use expression.Wrapper to encode a null column default. Check out how it's used in the resolve_defaults analyzer rule.

File: sql/analyzer/triggers.go [hunk 8]
```diff
@@ -355,49 +354,6 @@ func orderTriggersAndReverseAfter(triggers []*plan.CreateTrigger) []*plan.Create
 	return append(beforeTriggers, afterTriggers...)
 }
 
-func OrderTriggers(triggers []*plan.CreateTrigger) (beforeTriggers []*plan.CreateTrigger, afterTriggers []*plan.CreateTrigger) {
```
Comment (zachmu) [inline sql/analyzer/triggers.go:358]: You don't want to move this, I don't think. It's not DDL, it's about where to insert the trigger in the plan (before and after others)

**Review by zachmu — COMMENTED** [review]
This is definitely the right approach, just needs a little massaging to work

File: sql/information_schema/columns_table.go (hunk 6 above)
Comment (VinaiRachakonda) [reply sql/information_schema/columns_table.go:100]: I considered this but worried about inefficiencies. Expressions() gets called a lot so Iterating through all columns in database seems expensive.

File: sql/analyzer/triggers.go (hunk 8 above)
Comment (VinaiRachakonda) [reply sql/analyzer/triggers.go:358]: Yea i was problems with import loops. I don't think this method should be in the analyzer

File: sql/information_schema/columns_table.go (hunk 4 above)
Comment (VinaiRachakonda) [reply sql/information_schema/columns_table.go:41]: Won't this cause problems with resolution and bunch of analyzer logic? Essentially everywhere we are looking for a resolved table we are just getting a ColumnsTable. That's why I wanted this Node to wrap the ColumnTable

Comment (VinaiRachakonda) [reply sql/information_schema/columns_table.go:41]: Essentially we need to type the ColumnsTable in a ton of analyzer works. Wrapping a resolved table saves us a ton of potential edge cases IMO.

File: sql/information_schema/columns_table.go (hunk 6 above)
Comment (VinaiRachakonda) [reply sql/information_schema/columns_table.go:100]: This map is also essential in the resolve_column_defaults step where we need to actual column associated with the default to resolve. Moreover, we need the other columnNameToColumn map as a way to pass down the map down to the ultimate row iter

Comment (VinaiRachakonda) [reply sql/information_schema/columns_table.go:89]: nvm figured it out smh

File: sql/analyzer/resolve_tables.go [hunk 9]
```diff
@@ -126,8 +127,69 @@ func resolveTable(ctx *sql.Context, t *plan.UnresolvedTable, a *Analyzer) (sql.N
...
+				return false, nil
+			}
+
+			sct, ok := processNode.Child.(*plan.ShowCreateTable)
+			if !ok {
+				return false, nil
+			}
+
+			for _, col := range sct.GetTargetSchema() {
+				// Create a new column and update its name. This is useful for sorting later.
```
Comment (VinaiRachakonda) [inline sql/analyzer/resolve_tables.go:178]: Will fix this comment later.

File: sql/analyzer/resolve_tables.go [hunk 10]
```diff
@@ -126,8 +127,69 @@ func resolveTable(ctx *sql.Context, t *plan.UnresolvedTable, a *Analyzer) (sql.N
 		return handleTableLookupFailure(err, name, db, a, t)
 	}
 
+	resolvedTableNode := plan.NewResolvedTable(rt, database, nil)
+
+	// Check for the information_schema.columns table which needs to resolve all defaults
+	if database.Name() == "information_schema" && rt.Name() == "columns" {
```
Comment (zachmu) [inline sql/analyzer/resolve_tables.go:133]: Does this work case insensitive?

File: sql/analyzer/resolve_tables.go [hunk 11]
```diff
@@ -126,8 +127,69 @@ func resolveTable(ctx *sql.Context, t *plan.UnresolvedTable, a *Analyzer) (sql.N
...
+		return handleInfoSchemaColumnsTable(ctx, resolvedTableNode, a)
+	}
+
 	a.Log("table resolved: %s", t.Name())
-	return plan.NewResolvedTable(rt, database, nil), nil
+	return resolvedTableNode, nil
+}
+
+// handleInfoSchemaColumnsTable adds additional processing logic with the information_schema.columns table.
```
Comment (zachmu) [inline sql/analyzer/resolve_tables.go:141]: Be more specific, this comment doesn't tell me anything the name doesn't

File: sql/analyzer/resolve_tables.go [hunk 12]
```diff
@@ -126,8 +127,69 @@ func resolveTable(ctx *sql.Context, t *plan.UnresolvedTable, a *Analyzer) (sql.N
...
+	for _, db := range catalog.AllDatabases(ctx) {
+		err := sql.DBTableIter(ctx, db, func(t sql.Table) (cont bool, err error) {
+			// Construct a show create table node and analyze it to get a full resolved column default.
+			st := plan.NewShowCreateTable(plan.NewResolvedTable(t, db, nil), false)
+			analyzed, err := a.Analyze(ctx, st, nil)
+			if err != nil {
+				return false, err
+			}
+
+			processNode, ok := analyzed.(*plan.QueryProcess)
```
Comment (zachmu) [inline sql/analyzer/resolve_tables.go:167]: StripPassthroughNodes is your friend here

File: sql/analyzer/resolve_tables.go [hunk 13]
```diff
@@ -126,8 +127,69 @@ func resolveTable(ctx *sql.Context, t *plan.UnresolvedTable, a *Analyzer) (sql.N
...
+			if !ok {
+				return false, nil
+			}
+
+			sct, ok := processNode.Child.(*plan.ShowCreateTable)
+			if !ok {
+				return false, nil
```
Comment (zachmu) [inline sql/analyzer/resolve_tables.go:174]: Should be an error case here

File: sql/analyzer/triggers.go (hunk 8 above)
Comment (zachmu) [reply sql/analyzer/triggers.go:358]: It definitely belongs in the analyzer, where else would it go?

File: sql/analyzer/resolve_tables.go (hunk 9 above)
Comment (VinaiRachakonda) [reply sql/analyzer/resolve_tables.go:178]: After the review.

File: sql/information_schema/columns_table.go [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+// ColumnsTable describes the information_schema.columns table. It implements both sql.Node and sql.Table
+// as way to handle resolving column defaults.
+// Note: One way to approach this node, as it is some sort of analogue of sql.Table.
+type ColumnsTable struct {
+	// allColsWithDefaultValue is a list of all column with a non nill default value. Our goal is to resolve all of
```
Comment (zachmu) [inline sql/information_schema/columns_table.go:29]: comment about resolving doesn't seem accurate anymroe

File: sql/plan/ddl_trigger.go [hunk 15]
```diff
@@ -164,3 +166,48 @@ func (c *CreateTrigger) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, err
 		db: c.db,
 	}, nil
 }
+
+// OrderTriggers is a utility method that first sorts triggers into their precedence. It then splits the triggers into
+// before and after pairs.
+func OrderTriggers(triggers []*CreateTrigger) (beforeTriggers []*CreateTrigger, afterTriggers []*CreateTrigger) {
```
Comment (zachmu) [inline sql/plan/ddl_trigger.go:172]: Yeah definitely move this back, it's an analyzer function onlyu

**Review by zachmu — APPROVED** [review]
Generally looks fine, but move the trigger analysis code back

File: sql/plan/ddl_trigger.go (hunk 15 above)
Comment (VinaiRachakonda) [reply sql/plan/ddl_trigger.go:172]: Resolved in messages

**Pull request merged**